### ISSUE TITLE
PandasStore implementations can serialise unnecessarily twice certain types of dataframes

### ISSUE THREAD

**Issue by dimosped** [manahl/arctic#590]
#### Arctic Version

```
1.57.1
```

#### Arctic Store

```
VersionStore
```

#### Description of problem and/or code sample that reproduces the issue

The code which decides whether a given write handler can be used, does full-serialization to recarray of a dataframe whenever one or more object columns are detected.
https://github.com/manahl/arctic/blob/master/arctic/store/_pandas_ndarray_store.py#L172
https://github.com/manahl/arctic/blob/master/arctic/serialization/numpy_records.py#L140

This can happen in the following cases:
- Multi-indexed dataframe
- Multi-column dataframe
- any dataframe which container one or more columns of type object (includes string, or any column which might have even one row with a string/None value)

In the above cases, we end up serializing twice the dataframe, once for testing the write handler, and second to actually serialize when we call write/append.

For large data-frames, this slows down writes/appends and explodes further the memory usage.

**Comment by dimosped** [comment]
@bmoscon  @jamesblackburn @aflag  here is the draft PR for this fix.

It is being merged on my own arctic clone (onto the branch for the incremental serialization).
https://github.com/dimosped/arctic/tree/incremental-serialization-initial

After merging it will become automatically an extra change in this PR:
https://github.com/manahl/arctic/pull/578

I've made this optimisation optional (disabled by default) and added as thorough as possible tests.

**Comment by dimosped** [comment]
@aflag  this is interesting, we don't even manage to hit this check for n-dimensional DFs, as the serializer explodes:
https://github.com/manahl/arctic/blob/master/arctic/serialization/numpy_records.py#L151

```python
In[50]: import pandas as pd
   ...: import numpy as np
   ...: from arctic.serialization.numpy_records import DataFrameSerializer
   ...: 
   ...: dims = (2, 2, 2)
   ...: pdf = pd.DataFrame({'a': [np.zeros(dims), np.ones(dims)], 'b': [np.zeros(dims), np.ones(dims)]})
   ...: # recs = pdf.to_records()
   ...: f_serializer = DataFrameSerializer()
   ...: recs = df_serializer.serialize(pdf)
Traceback (most recent call last):
  File ".egg/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-50-6833e8b9137b>", line 9, in <module>
    recs = df_serializer.serialize(pdf)
  File "arctic/arctic/serialization/numpy_records.py", line 275, in serialize
    return self._to_records(item, string_max_len, forced_dtype)
  File "arctic/arctic/serialization/numpy_records.py", line 144, in _to_records
    rtn = np.rec.fromarrays(arrays, dtype=dtype, names=names)
  File "numpy-1.9.2-py2.7-linux-x86_64.egg/numpy/core/records.py", line 558, in fromarrays
    raise ValueError("array-shape mismatch in array %d" % k)
ValueError: array-shape mismatch in array 1
```

The same error occurs both for numpy/pandas 1.9.2/0.16.2  and 1.15.0/0.23.3 

Nonetheless I am enhancing the tests with ndimensional for compatibility regression here:
https://github.com/dimosped/arctic/pull/1

**Comment by bmoscon** [comment]
was this ever merged in?

**Comment by dimosped** [comment]
merged on the main incremental serialization branch
https://github.com/dimosped/arctic/pull/1/files  

This will make it to master and be released after 1.68.1 hotfix
